# Credit Risk Classification

Credit risk poses a classification problem that’s inherently imbalanced. This is because healthy loans easily outnumber risky loans. In this Challenge, you’ll use various techniques to train and evaluate models with imbalanced classes. You’ll use a dataset of historical lending activity from a peer-to-peer lending services company to build a model that can identify the creditworthiness of borrowers.

## Instructions:

This challenge consists of the following subsections:

* Split the Data into Training and Testing Sets

* Create a Logistic Regression Model with the Original Data

* Predict a Logistic Regression Model with Resampled Training Data 

### Split the Data into Training and Testing Sets

Open the starter code notebook and then use it to complete the following steps.

1. Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

2. Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

    > **Note** A value of `0` in the “loan_status” column means that the loan is healthy. A value of `1` means that the loan has a high risk of defaulting.  

3. Check the balance of the labels variable (`y`) by using the `value_counts` function.

4. Split the data into training and testing datasets by using `train_test_split`.

### Create a Logistic Regression Model with the Original Data

Employ your knowledge of logistic regression to complete the following steps:

1. Fit a logistic regression model by using the training data (`X_train` and `y_train`).

2. Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

3. Evaluate the model’s performance by doing the following:

    * Calculate the accuracy score of the model.

    * Generate a confusion matrix.

    * Print the classification report.

4. Answer the following question: How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

### Predict a Logistic Regression Model with Resampled Training Data

Did you notice the small number of high-risk loan labels? Perhaps, a model that uses resampled data will perform better. You’ll thus resample the training data and then reevaluate the model. Specifically, you’ll use `RandomOverSampler`.

To do so, complete the following steps:

1. Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points. 

2. Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

3. Evaluate the model’s performance by doing the following:

    * Calculate the accuracy score of the model.

    * Generate a confusion matrix.

    * Print the classification report.
    
4. Answer the following question: How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

### Write a Credit Risk Analysis Report

For this section, you’ll write a brief report that includes a summary and an analysis of the performance of both machine learning models that you used in this challenge. You should write this report as the `README.md` file included in your GitHub repository.

Structure your report by using the report template that `Starter_Code.zip` includes, and make sure that it contains the following:

1. An overview of the analysis: Explain the purpose of this analysis.


2. The results: Using bulleted lists, describe the balanced accuracy scores and the precision and recall scores of both machine learning models.

3. A summary: Summarize the results from the machine learning models. Compare the two versions of the dataset predictions. Include your recommendation for the model to use, if any, on the original vs. the resampled data. If you don’t recommend either model, justify your reasoning.

In [1]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import balanced_accuracy_score
from sklearn.metrics import confusion_matrix
from imblearn.metrics import classification_report_imbalanced
# Import the train_test_learn module
from sklearn.model_selection import train_test_split
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression
# Import the RandomOverSampler module form imbalanced-learn
from imblearn.over_sampling import RandomOverSampler

import warnings
warnings.filterwarnings('ignore')

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [2]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
lending_data_df = pd.read_csv(Path("./Resources/lending_data.csv"))

# Review the DataFrame
lending_data_df.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [3]:
# Separate the data into labels and features

# Separate the y variable, the labels
y = lending_data_df["loan_status"]

# Separate the X variable, the features
X = lending_data_df.drop(columns="loan_status")

In [4]:
# Review the y variable Series
y.head()

0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64

In [5]:
# Review the X variable DataFrame
X.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000

### Step 3: Check the balance of the labels variable (`y`) by using the `value_counts` function.

In [6]:
# Check the balance of our target values
y.value_counts()

0    75036
1     2500
Name: loan_status, dtype: int64

### Step 4: Split the data into training and testing datasets by using `train_test_split`.

In [7]:
# Split the data using train_test_split
# Assign a random_state of 1 to the function
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state = 1)

In [25]:
# Calculate the training stats and print them
# calculate the healthy loans
healthy_loans_train = y_train.value_counts()[0]
# calculate the high-risk loans
high_risk_loans_train = y_train.value_counts()[1]
# calculate the total loans
total_loans_train = y_train.value_counts().sum()
# calculate the healthy loans percentage
healthy_loans_percentage_train = (healthy_loans_train / total_loans_train) * 100
# calculate the high-risk loans percentage
high_risk_loans_percentage_train = (high_risk_loans_train / total_loans_train) * 100

print(f"Healthy loans are {healthy_loans_train} out of {total_loans_train} - {healthy_loans_percentage_train:.2f}%") 
print(f"High Risk loans are {high_risk_loans_train} out of {total_loans_train} - {high_risk_loans_percentage_train:.2f}%")

Healthy loans are 56271 out of 58152 - 96.77%
High Risk loans are 1881 out of 58152 - 3.23%


---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [8]:
# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
logistic_regression_model = LogisticRegression(random_state = 1)

# Fit the model using training data
logistic_regression_model.fit(X_train, y_train)

LogisticRegression(random_state=1)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [9]:
# Make a prediction using the testing data
predictions = logistic_regression_model.predict(X_test)

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [10]:
# Print the balanced_accuracy score of the model
balanced_accuracy_score(y_test, predictions)

0.9520479254722232

In [11]:
# Generate a confusion matrix for the model
print(confusion_matrix(y_test, predictions))

[[18663   102]
 [   56   563]]


In [12]:
# Print the classification report for the model
print(classification_report_imbalanced(y_test, predictions))

                   pre       rec       spe        f1       geo       iba       sup

          0       1.00      0.99      0.91      1.00      0.95      0.91     18765
          1       0.85      0.91      0.99      0.88      0.95      0.90       619

avg / total       0.99      0.99      0.91      0.99      0.95      0.91     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** Precision for healthy loan (0) is 1 - Of all the health loans predicted by the Logistic Regression, almost all of them are actually healthy.

Recall for health loan (0) is 0.99 - 99% of the actual healthy loans is predicted fine and only 1% of them are inaccurately predicted as high-risk loans. The logistic regression is able to classify the healthy loans without any issues. 

This is expected because there are 56271 healthy loans out of the total 58152 loans that are used for the training. Healthy loans are almost 97% of the total loans in the training data. So, logistic regression is well exposed as far as training on healthy loans is concerned.  

Precision for high-risk loan (1) is 0.85 - Of all the high-risk loan predicted by the Logistic Regression, 85% of them are actually high-risk loans and 15% of them are healthy loans. There is probably a small penalty in terms of the extra scrutiny of these 15% health loans before they are approved if there is extra processing involved for high risk loans otherwise we have some unhappy customers who should have been approved.

Recall of the high-risk loans (1) is 0.91 - 91% of the actual high risk loans are predicted fine and 9% of the actual high risk loans are classified as healthy loans. This could be a problem for a company trying to accurately predict the high risk loans because the penalty/cost of default is very high on the high risk loans.

High Risk loans are 1881 out of total 58152 loans in the training set. This is 3.23% of the total loans. The High Risk loans are under represented in the training data and therefore the logistic regression is not well trained on the high risk loans. 

One way to fix this is to increase the percentage of the high risk loans in the data set so that they match the health loans either through oversampling of the high risk loans or under sampling of the health loans.

---

## Predict a Logistic Regression Model with Resampled Training Data

### Step 1: Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points. 

In [13]:
# Instantiate the random oversampler model
# # Assign a random_state parameter of 1 to the model
random_over_sampler = RandomOverSampler(random_state = 1)

# Fit the original training data to the random_oversampler model
X_oversampled, y_oversampled = random_over_sampler.fit_resample(X_train, y_train)

In [14]:
# Count the distinct values of the resampled labels data
y_oversampled.value_counts()

0    56271
1    56271
Name: loan_status, dtype: int64

### Step 2: Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

In [15]:
# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
logistic_regression_model_ros = LogisticRegression(random_state = 1)

# Fit the model using the resampled training data
logistic_regression_model_ros = logistic_regression_model_ros.fit(X_oversampled, y_oversampled)

# Make a prediction using the testing data
y_pred_oversampled = logistic_regression_model_ros.predict(X_test)

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [16]:
# Print the balanced_accuracy score of the model 
balanced_accuracy_score(y_test, y_pred_oversampled)

0.9936781215845847

In [17]:
# Generate a confusion matrix for the model
print(confusion_matrix(y_test, y_pred_oversampled))

[[18649   116]
 [    4   615]]


In [18]:
# Print the classification report for the model
print(classification_report_imbalanced(y_test, y_pred_oversampled))

                   pre       rec       spe        f1       geo       iba       sup

          0       1.00      0.99      0.99      1.00      0.99      0.99     18765
          1       0.84      0.99      0.99      0.91      0.99      0.99       619

avg / total       0.99      0.99      0.99      0.99      0.99      0.99     19384



### Step 4: Answer the following question

**Question:** How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** Precision for healthy loan (0) is 1 - Of all the health loans predicted by the Logistic Regression, almost all of them are actually healthy.

Recall for health loan (0) is 0.99 - 99% of the actual healthy loans are predicted fine and only 1% of them are inaccurately predicted as high-risk loans. The logistic regression is able to classify the healthy loans without any issues. 

This didn't change from the original training of data because we still have the same number of healthy loans (56271) and they were well represented in the original regression's training. 

Precision for high-risk loan (1) is 0.84 - Of all the high-risk loan predicted by the Logistic Regression, 84% of them are actually high-risk loans and 16% of them are healthy loans. There is probably a small penalty in terms of the extra scrutiny of these 16% health loans before they are approved if there is extra processing involved for high risk loans otherwise we have some unhappy customers (with some cost and loss) who should have been approved. As compared to the original regression, there is not much difference in terms of precision for high-risk loan.

Recall of the high-risk loans (1) is 0.99 - 99% of the actual high risk loans are predicted accurately and only 1% of the actual high risk loans are classified as healthy loans. This is a huge improvement from the original training when we were incorrectly predicted 9% of the high risk loans as healthy loans. So, we have reduced our penalty/cost of default associated with approving high risk loans as health loans.

Random Oversampling of the high-risk training data has helped us improve the recall of the high-risk loans and drastically reduced the penalty/cost of default of approving high risk loans as health loans. 

We could potentially improve the precision also with synthetic sampling and collection of more high risk loan data.